File: 0438-find-all-anagrams-in-a-string/0438-find-all-anagrams-in-a-string.py
class Solution(object):
    def findAnagrams(self, s, p):
        """
        :type s: str
        :type p: str
        :rtype: List[int]
        """
        freq_p={}
        for c in p:
            freq_p[c]=freq_p.get(c,0)+1
        window={}
        j=0
        ans=[]
        n=len(p)
        for i in range(len(s)):
            if s[i] in freq_p:
                window[s[i]]=window.get(s[i],0)+1
            else:
                window.clear()
                j=i
            if i-j+1==n:
                if window==freq_p:
                    ans.append(j)
                if s[j] in window:
                    window[s[j]]-=1
                    if window[s[j]]==0:
                        del window[s[j]]
                j+=1
        return ans
            
                

        

File: 0438-find-all-anagrams-in-a-string/0438-find-all-anagrams-in-a-string.cpp
class Solution {
public:
    vector<int> findAnagrams(string s, string p) {
        int n=s.size(),j=0,count=0;
        vector<int> res;
        unordered_map<char,int> pw,sw;
        for(char c:p) pw[c]++;
        for(int i=0;i<n;i++){
            if(pw.count(s[i])){
                sw[s[i]]++;
            }else{
                j=i;
                sw.clear();
            }
            if(i-j+1==p.size()){
                if(sw==pw)res.push_back(j);
                if(sw.count(s[j])){
                    sw[s[j]]--;
                    if(sw[s[j]]==0){
                        sw.erase(s[j]);
                    }
                }
                j++;
            }

        }
        return res;

    }
};